doc/proposals/scorecard-user-tests.md

```diff
@@ -0,0 +1,153 @@
+type UserDefinedTest struct {
+    // Path to cr to be used for testing
+    CRPath string `mapstructure:"cr"`
+    // Expected resources and status
+    Expected Expected `mapstructure:"expected"`
+    // Sub-tests modifying a few fields with expected changes
+    Modifications []Modification `mapstructure:"modifications"`
+}
+
+type Expected struct {
+	// Resources expected to be created after the operator reacts to the CR
+	Resources []map[string]interface{} `mapstructure:"resources"`
+	// Expected values in CR's status after the operator reacts to the CR
+	Status map[string]interface{} `mapstructure:"status"`
+}
+
+// Modifications specifies a spec field to change in the CR with the expected results
+type Modification struct {
+    // a map of the spec fields to modify
+    Spec map[string]interface{} `mapstructure:"spec"`
+    // Expected resources and status
+    Expected Expected `mapstructure:"expected"`
+}
+```
+
+For `Status` and `Resources` fields, we can implement a bit of extra computation instead of simple string checking. For instance,
+in the memcached-operator test, we should expect that the length of the `nodes` field (which is an array) has a certain length. To implement functions like
+these, we can create some functions for these checks that are prepended by `scorecard_function_` and take an array of objects. For instance, in the above
+example, `scorecard_function_length` would check that each field listed under it matches the specified length (like `nodes: 4`). If the yaml key does not
+start with `scorecard_function_`, we do a simple match (like `status/readyReplicas: 4`).
+
+This design would allow us to replace the old "Operator actions are reflected in status" (which would be tested by the `expected/status` check) and
+"Writing into CRs has an effect" (which would be tested by the `expected/resources` check) tests.
+
+### Plugin System
+
+In order to increase the flexibility of the user defined tests and allow users to implement more complex E2E style tests for scorecard,
+the user-defined tests will be implemented via a plugin system. Users will specify the path of the script they wish to run as well
+as environment variable to set for the command. The command would then print out the result as JSON to stdout. Here is an example:
+
+```yaml
+user_defined_tests:
+- path: "scorecard/simple-scorecard.sh"
+  env:
+    - CONFIG_FILE: "scorecard/simple-scorecard.yaml"
+- path: "scorecard/go-test.sh"
+  env:
+    - ENABLE_SCORECARD: true
+    - NAMESPACED_MANIFEST: "deploy/namespaced_init.yaml"
+    - GO_TEST_FLAGS: "-parallel=1"
+```
+
+The above is an example of a user-defined test suite with 2 tests: the new simple scorecard tests described above and a test
+built using the Operator SDK's test framework that's been modified to be able to output results in the standardized JSON output.
+
+Below is an example of what the JSON output of a test would look like:
+
+```json
```

Comment:
I would like this to more certain things

1. hard failures. Did the test fail in a way that nothing else should be considered
2. I assume that once plugin could have more than 1 test. I would like to see a summary struct and a list of test results for the individual runs. I think this will be more flexible. 

I would also like this to be the output of the scorecard test. I prefer computer readable structured data to the current output. Maybe have a `-h` option that prints out as a human-readable format, but the default is a computer readable makes more sense to me.

Reply:
For 1: What would be a good way to handle that? Add another field to the JSON output? Or maybe check the return code (if command returned 0 use JSON output; if command returned not 0, assume hard fail)?

For 2: In this example JSON I'm showing 2 tests from a single plugin: "actions_reflected_in_status" and "my_custom_tests".

I think someone was asking if we could have JSON output for the scorecard itself. I thought I made an issue for that, but it looks like I didn't. I'll create a new issue to track that. It'll probably be pretty much the same output as the example plugin output I have here (the plugin output example is just a JSONified array of TestResult objects)

Reply:
I would consider a different structure. I think that something that gives someone reading this output easy path to see if everything is good, but also can deep dive into the results if something goes wrong.

Reply:
So to recap for the updated proposal: To indicate a hard failure(e.g error associate with `t.Fatal()`) the test result would have `state: error` and some error message appended to `errors`.

And if it's just regular errors like with `t.Error()` the output of a test result would be `state: failed` or `state: partial_pass` and error messages appended to `errors`.

@shawn-hurley Do you think that's clear enough to distinguish between a fatal/hard failure vs regular errors?

Reply:
Yeah, that makes sense for the fatal/hard failure.

I was also thinking that you would want to give the denominator. Like the number of tests run or possible points? I would also suggest that you note the total_score is a percentage. that is not super clear.

@brianwcook can you have your team take a look at this format to make sure that it works for your team?

Reply:
@shawn-hurley Updated proposal with your suggestions.

Reply:
I think we need to also account for the possibility that the plugin executable exits with a non-zero return code and/or with output that is not what we expect. And describe how operator-sdk handles that situation.